File: backend/unimate/src/main/java/com/unimate/domain/user/user/service/UserAuthService.java
package com.unimate.domain.user.user.service;

import com.unimate.domain.user.user.dto.UserLoginRequest;
import com.unimate.domain.user.user.dto.UserSignupRequest;
import com.unimate.domain.user.user.dto.UserSignupResponse;
import com.unimate.domain.user.user.entity.User;
import com.unimate.domain.user.user.repository.UserRepository;
import com.unimate.domain.verification.service.VerificationService;
import com.unimate.global.auth.dto.Tokens;
import com.unimate.global.auth.model.SubjectType;
import com.unimate.global.auth.service.TokenService;
import com.unimate.global.exception.ServiceException;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class UserAuthService {

    private final UserRepository userRepository;
    private final BCryptPasswordEncoder passwordEncoder;
    private final TokenService tokenService;
    private final VerificationService verificationService;


    @Transactional
    public UserSignupResponse signup(UserSignupRequest req) {
        if (userRepository.existsByEmail(req.getEmail())) {
            throw ServiceException.badRequest("이미 가입된 이메일입니다.");
        }

        verificationService.assertVerifiedEmailOrThrow(req.getEmail());

        User user = new User(
                req.getName(),
                req.getEmail(),
                passwordEncoder.encode(req.getPassword()),
                req.getGender(),
                req.getBirthDate(),
                req.getUniversity()
        );
        user.verifyStudent();

        User savedUser = userRepository.save(user);
        verificationService.consumeVerification(req.getEmail());

        return new UserSignupResponse(savedUser.getId(), savedUser.getEmail(), savedUser.getName());
    }

    @Transactional
    public Tokens login(UserLoginRequest req) {
        User user = userRepository.findByEmail(req.getEmail())
                .orElseThrow(() -> ServiceException.notFound("이메일을 찾을 수 없습니다."));

        if (!passwordEncoder.matches(req.getPassword(), user.getPassword())) {
            throw ServiceException.unauthorized("비밀번호가 일치하지 않습니다.");
        }

        return tokenService.issueTokens(SubjectType.USER, user.getId(), user.getEmail());
    }

    @Transactional(readOnly = true)
    public String reissueAccessToken(String refreshToken) {
        return tokenService.reissueAccessToken(refreshToken);
    }

    @Transactional
    public void logout(String refreshToken) {
        tokenService.logout(refreshToken);
    }
}

File: backend/unimate/src/main/java/com/unimate/domain/user/user/service/UserAuthService.kt
package com.unimate.domain.user.user.service

import com.unimate.domain.user.user.dto.UserLoginRequest
import com.unimate.domain.user.user.dto.UserSignupRequest
import com.unimate.domain.user.user.dto.UserSignupResponse
import com.unimate.domain.user.user.entity.User
import com.unimate.domain.user.user.repository.UserRepository
import com.unimate.domain.verification.service.VerificationService
import com.unimate.global.auth.dto.Tokens
import com.unimate.global.auth.model.SubjectType
import com.unimate.global.auth.service.TokenService
import com.unimate.global.exception.ServiceException
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.function.Supplier

@Service
class UserAuthService(
    private val userRepository: UserRepository,
    private val passwordEncoder: BCryptPasswordEncoder,
    private val tokenService: TokenService,
    private val verificationService: VerificationService
) {

    @Transactional
    fun signup(req: UserSignupRequest): UserSignupResponse {
        if (userRepository.existsByEmail(req.email)) {
            throw ServiceException.conflict("이미 가입된 이메일입니다.")
        }

        verificationService.assertVerifiedEmailOrThrow(req.email)

        val user = User(
            req.name,
            req.email,
            passwordEncoder.encode(req.password),
            req.gender,
            req.birthDate,
            req.university
        )
        user.verifyStudent()

        val savedUser = userRepository.save(user)
        verificationService.consumeVerification(req.email)

        val userId = requireNotNull(savedUser.id) { "저장된 사용자의 ID가 null입니다." }

        return UserSignupResponse(userId, savedUser.email, savedUser.name)
    }

    @Transactional
    fun login(req: UserLoginRequest): Tokens {
        val user = userRepository.findByEmail(req.email)
            .orElseThrow<ServiceException?>(Supplier { ServiceException.unauthorized("이메일 또는 비밀번호가 일치하지 않습니다.") })

        if (!passwordEncoder.matches(req.password, user.password)) {
            throw ServiceException.unauthorized("이메일 또는 비밀번호가 일치하지 않습니다.")
        }

        return tokenService.issueTokens(SubjectType.USER, user.id!!, user.email)
    }

    @Transactional(readOnly = true)
    fun reissueAccessToken(refreshToken: String): String {
        return tokenService.reissueAccessToken(refreshToken)
    }

    @Transactional
    fun logout(refreshToken: String) {
        tokenService.logout(refreshToken)
    }
}